### DISCUSSION TITLE
Live Chord Input from Guitar Hero Controller

### DISCUSSION THREAD

**Discussion by deleolajide** [jjazzboss/JJazzLab#201] [Ideas]
I have been making music for a while creating live backing music using a Guitar Hero controller with the Real Guitar software and a [custom Midi controller I developed myself](https://github.com/deleolajide/peakcontroller) to control arranger modules like the Yahama QY100 or the Ketron SD90

I was thinking of converting JJazzLab-X into an live arranger that I could control from my Guitar Hero controller or from a cheap wireless numeric keypad that would offer me more control beyond the 5 buttons limitation I currently have.

I noticed that is is a [Midi Input branch](https://github.com/jjazzboss/JJazzLab-X/tree/MidiInput).

Is it useable? How far did you get with it?

**Comment by jjazzboss** [comment]
This Midi Input branch is really not finished, goal was to record a melody or solo. On the master branch Midi input is now enabled, I added a feature to control play/pause/stop commands from a Midi keyboard (see https://github.com/jjazzboss/JJazzLab-X/issues/161).

As of today JJazzLab is not designed to work as a realtime arranger, i.e. capture chord in real time and generate music on the fly. It's perfectly doable, but not a simple task.

**Reply by deleolajide** [reply]
I am hoping to achieve something similar to the launchpad feature of Ketron SD9 series. It is a hybrid of a real-time arranger and a backing track generator. I am trying to create a live backing track generator that has varying levels of complexity depending on what type of input device is controlling it. The devices will vary from USB/wireless numeric keypad to game controllers like my Guitar Hero game controller

In the simplest mode, a user can use buttons to start, pause/resume and stop a complete song. 

Next step up would be to add more buttons that will enable the ability to jump between or repeat song sections playing the pre-recorded chord sequences like using a sampler for live performance.

The most flexible mode would be the ability to override the pre-recorded chord sequence of a song section with live chords from yet more buttons on the input device. At the moment, I can play about 12+ chords with the 5 buttons on the Guitar Hero. If no chord buttons are pressed, the sequence continues with the pre-recorded sequence by default.

In practice, it means that I can take a very familiar song, play melody as well as control the virtual backing band live in real-time. I can also have the ability to handle a more complex song, concentrating on the melody while the virtual backing band plays what was practised before hand. The melody will be either direct singing or a voice-to-midi controller.

**Reply by deleolajide** [reply]
> I am hoping to achieve something similar to the launchpad feature of Ketron SD9 series. It is a hybrid of a real-time arranger and a backing track generator. I am trying to create a live backing track generator that has varying levels of complexity depending on what type of input device is controlling it. The devices will vary from USB/wireless numeric keypad to game controllers like my Guitar Hero game controller
> 
> In the simplest mode, a user can use buttons to start, pause/resume and stop a complete song.
> 
> Next step up would be to add more buttons that will enable the ability to jump between or repeat song sections playing the pre-recorded chord sequences like using a sampler for live performance.
> 
> The most flexible mode would be the ability to override the pre-recorded chord sequence of a song section with live chords from yet more buttons on the input device. At the moment, I can play about 12+ chords with the 5 buttons on the Guitar Hero. If no chord buttons are pressed, the sequence continues with the pre-recorded sequence by default.
> 
> In practice, it means that I can take a very familiar song, play melody as well as control the virtual backing band live in real-time. I can also have the ability to handle a more complex song, concentrating on the melody while the virtual backing band plays what was practised before hand. The melody will be either direct singing or a voice-to-midi controller.

I came across this while searching for something else and pleasantly surprised myself. For the sake of posterity, I did start this journey and actively implementing what I described here in JavaScript using WebAudio and WebMidi in a project called [OrinAyo](https://github.com/Jus-Be/orin-ayo)

**Comment by deleolajide** [comment]
>It's perfectly doable, but not a simple task.

Could you please provide any hints about how the code can be adapted to do this. Thanks 👍

**Comment by jjazzboss** [comment]
I updated today the Developer's guide (new Getting started page): https://jjazzlab.gitbook.io/developer-guide/
To start you should have a look at this class, which implements the remote play/pause commands via incoming Midi notes.
https://github.com/jjazzboss/JJazzLab-X/blob/master/MusicControlActions/src/org/jjazz/ui/musiccontrolactions/RemoteController.java
